src/graph/transform/cuda/cuda_to_block.cu

```diff
@@ -0,0 +1,972 @@
+  }
+
+
+  Mapping<IdType> * LhsHashTable(
+      const size_t index) {
+    return HashTable(index);
+  }
+
+  Mapping<IdType> * RhsHashTable(
+      const size_t index) {
+    return HashTable(index+rhs_offset_);
+  }
+
+  const Mapping<IdType> * LhsHashTable(
+      const size_t index) const {
+    return HashTable(index);
+  }
+
+  const Mapping<IdType> * RhsHashTable(
+      const size_t index) const {
+    return HashTable(index+rhs_offset_);
+  }
+
+  IdType LhsHashSize(
+      const size_t index) const {
+    return HashSize(index);
+  }
+
+  IdType RhsHashSize(
+      const size_t index) const {
+    return HashSize(rhs_offset_+index);
+  }
+
+  size_t Size() const {
+    return masks_.size();
+  }
+
+ private:
+  size_t num_types_;
+  size_t rhs_offset_;
+  std::vector<int> masks_;
+  std::vector<size_t> hash_prefix_;
+  Mapping<IdType> * hash_tables_;
+  DGLContext ctx_;
+
+  static size_t GetHashSize(const size_t num_nodes) {
+    const size_t num_bits = static_cast<size_t>(
+        std::ceil(std::log2(static_cast<double>(num_nodes))));
+    const size_t next_pow2 = 1ull << num_bits;
+
+    // make sure how hash size is not a power of 2
+    const size_t hash_size = (next_pow2 << HASH_SCALING)-1;
+
+    return hash_size;
+  }
+
+  inline Mapping<IdType> * HashTable(
+      const size_t index) {
+    return hash_tables_ + hash_prefix_[index];
+  }
+
+  inline const Mapping<IdType> * HashTable(
+      const size_t index) const {
+    return hash_tables_ + hash_prefix_[index];
+  }
+
+  inline IdType HashSize(
+      const size_t index) const {
+    return hash_prefix_[index+1]-hash_prefix_[index];
+  }
+};
+
+template<typename IdType>
+class DeviceNodeMapMaker {
+ public:
+  constexpr static const int BLOCK_SIZE = 256;
+  constexpr static const size_t TILE_SIZE = 1024;
+
+  static size_t RequiredPrefixSumBytes(
+      const size_t max_num) {
+    size_t prefix_sum_bytes;
+    CUDA_CALL(cub::DeviceScan::ExclusiveSum(
+          nullptr,
+          prefix_sum_bytes,
+          static_cast<IdType*>(nullptr),
+          static_cast<IdType*>(nullptr),
+          RoundUp(max_num, TILE_SIZE)+1));
+    return RoundUp(prefix_sum_bytes, 8);
+  }
+
+  DeviceNodeMapMaker(
+      const std::vector<int64_t> maxNodesPerType) :
+      max_num_nodes_(0),
+      workspace_size_bytes_(0) {
+    max_num_nodes_ = *std::max_element(maxNodesPerType.begin(),
+        maxNodesPerType.end());
+
+    workspace_size_bytes_ = RequiredPrefixSumBytes(max_num_nodes_) +
+        sizeof(IdType)*(max_num_nodes_+1);
+  }
+
+  size_t RequiredWorkspaceBytes() const {
+    return workspace_size_bytes_;
+  }
+
+  /**
+  * @brief This function builds node maps for each node type, preserving the
+  * order of the input nodes.
+  *
+  * @param nodes The set of input nodes.
+  * @param nodes_maps The constructed node maps.
+  * @param workspace The scratch space to use.
+  * @param workspaceSize The size of the scratch space.
+  * @param stream The stream to operate on.
```

Comment:
The docstring misses some of the arguments.

Reply:
Updated in 
85196a2